libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * UnusedFunctionParameterPruner: Optimiser step that removes unused-parameters function.
+ */
+
+#include <libyul/optimiser/UnusedFunctionParameterPruner.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/Visitor.h>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+using namespace solidity::langutil;
+
+namespace
+{
+
+/**
+ * First step of UnusedFunctionParameterPruner: Find functions with whose parameters are not used in
+ * its body.
+ */
+struct FindFunctionsWithUnusedParameters
+{
+	void operator()(FunctionDefinition const& _function)
+	{
+		map<YulString, size_t> namesFound = ReferencesCounter::countReferences(_function.body);
+
+		// We skip if the function body
+		// 1. is empty
+		// 2. is a single statement, that is an assignment statement with RHS being a function call
+		// 3. is a single expression-statement, that is a function call
```

Comment:
```suggestion
		// 3. is a single expression-statement, that is a function call
		// This allows this optimizer step to run again on its output without making things worse.
```

Reply:
Or maybe something like: The above cases are simple enough such that the inliner alone can remove the parameters.